### PR TITLE
[SECURITY-3033] Submit POST requests as needed

### PR DESCRIPTION
**Pull request by daniel-beck**
This is an attempt to resolve [SECURITY-3033](https://www.jenkins.io/security/advisory/2023-07-12/#SECURITY-3033).

It's a lot more complex than the usual solution of adding `@POST` or `@RequirePOST` to a web method and calling it a day.

First, the sidebar link directly intiates the action, so this needed a `task.jelly` for that that sets the `<l:task post="true"/>` attribute.

The next problem is that `RebuildAction#doIndex` does everything (meaning, triggering a build or redirecting to the form to set parameters), depending on the configuration. Adding `@RequirePOST` or `@POST` would secure the part that triggers a build, but `<l:task post="true" …>` doesn't redirect to the `Location` returned from the response, so if a form needs to be filled, that would never happen if `post="true"` is set unconditionally.

Additionally, `#getUrlName` cannot end in a slash, but an `<l:task href="${action.urlName}" post="true" … />` doesn't work, as the Stapler-internal forward from `…/rebuild` to `…/rebuild/` loses the HTTP POST method. So introduce `#getTaskUrl` and link there instead.

So the `task.jelly` needed a way to determine whether to send a POST request or not, beforehand. This was straightforward enough for the regular `RebuildAction` (for builds), but the job-scoped `RebuildLastCompletedBuildAction` could not call into `RebuildAction` after looking it up on the build, since everything was operating on Stapler ancestors -- and there's no build ancestor when on the project page.

So, some cleanup: `RebuildAction` is no longer a superclass of `RebuildLastCompletedBuildAction`, both now share the common ancestor `AbstractRebuildAction`, which provides the new `task.jelly` shared between the two. Make `RebuildAction` a `RunAction2` (for backwards compatibility when `RebuildAction` was persisted with builds, newly removed here) and provide the `Run` to it. Remove all code from `Rebuilder` since everything is transient now; instead have `RebuildActionFactory` (no longer a `TransientBuildActionFactory`) check whether a persisted action already exists before attaching a transient one.

This PR also deletes a bunch of code that appears long obsolete. I urge maintainers tempted to merge this to double-check; I am unaware why this code has been kept around for many years and there may be a good reason to have done this.

Jobs without builds now no longer allow "Rebuild Last"), because that seemed weird to work in the first place.

### Testing done

Configured three freestyle projects (no params, params, and params with auto-rebuild) and clicked the actions. They seemed about correct.

I have tested with the persisted action a bit (rebuild of historical builds both with and without parameters, checking there's no second action), but not extensively.

Additionally, if someone else wants to take this on, I strongly recommend adding tests with historical build records with persisted `RebuildAction`, as well as adding tests confirming the effectiveness of the security fix (basically, what the tests kind of did until they were adapted, except with `assertThrows` so they pass). Any obscure undocumented use cases (e.g., "expect reasonable outcome when someone navigates to `…/rebuild` directly and a parameters page should show up") will need to have tests added as well.

```[tasklist]
### Submitter checklist
- [ ] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [ ] Ensure that the pull request title represents the desired changelog entry
- [ ] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

**Comment by asafnahshon** [comment]
Maintainers are gone, eh? what should we do then?

Do you need more manual testing?

**Comment by morganchristiansson** [comment]
Last PR that was merged has [this comment](https://github.com/jenkinsci/rebuild-plugin/pull/147#issuecomment-1585858693)

CC @GLundh

**Comment by daniel-beck** [comment]
Note that in https://github.com/jenkins-infra/update-center2/commit/fcb1ae7dd1b0ea156e7510073b5d7e52b9fe01f5#r121610706 the maintainer informed us they have no time to maintain the plugin anymore.

As a user, you could download and install [the CI build ("incremental")](https://repo.jenkins-ci.org/incrementals/com/sonyericsson/hudson/plugins/rebuild/rebuild/325.v5ee8da_5293dc/) to address the issue (note that the warning shown on the UI needs to be dismissed manually).

As a Java developer, consider [becoming a maintainer of this plugin](https://www.jenkins.io/doc/developer/plugin-governance/adopt-a-plugin/).

**Comment by asafnahshon** [comment]
Downloaded and installed the "incremental" build as Daniel suggested, works fine in production, thank you.

**Comment by daniel-beck** [comment]
Oversight when I wrote the custom `action.jelly`, should be fixed now. (FTR I'm unable to do what `t:actions` does and use `Functions#getActionUrl`, because not using `Action#getUrlName` is the entire point of the custom `action.jelly`.)

File: src/test/java/com/sonyericsson/rebuild/RebuildValidatorTest.java [hunk 1]
```diff
@@ -258,20 +267,20 @@ public void testWhenProjectWithParamsThenRebuildProjectExecutesRebuildOfLastBuil
...
         j.submit(rebuildConfigPage.getFormByName("config"));
 
         HtmlPage projectPage = j.createWebClient().getPage(project);
         WebAssert.assertLinkPresentWithText(projectPage, "Rebuild Last");
 
         HtmlAnchor rebuildHref = projectPage.getAnchorByText("Rebuild Last");
-        assertEquals("Rebuild Last should point to the second build", "/jenkins/"
-                + project.getUrl() + "lastCompletedBuild/rebuild",
+        assertEquals("Rebuild Last should point to the second build",
+                "lastCompletedBuild/rebuild/parameterized",
```
Comment (Evan-Bluhm) [inline src/test/java/com/sonyericsson/rebuild/RebuildValidatorTest.java:283]: Test was actually okay before
```suggestion
        assertEquals("Rebuild Last should point to the second build",
                j.contextPath + "/" + project.getUrl() + "lastCompletedBuild/rebuild/parameterized",
```

Comment (daniel-beck) [reply src/test/java/com/sonyericsson/rebuild/RebuildValidatorTest.java:283]: Thanks for the reminder.

**Comment by lmgray** [comment]
do we want to link [https://issues.jenkins.io/browse/JENKINS-71932](https://issues.jenkins.io/browse/JENKINS-71932) and assign to you @daniel-beck ?

**Comment by daniel-beck** [comment]
I'm not sure filing an issue for an open PR makes sense. Who's the audience of that? As I wrote further up, I'm not interested in maintaining this plugin, so this PR should be considered unsupported, and I'm certainly not keeping an eye on newly reported issues; and any maintainers aren't responsible for my mistakes (at least until they merge them).

**Comment by lmgray** [comment]
> I'm not sure filing an issue for an open PR makes sense. Who's the audience of that? As I wrote further up, I'm not interested in maintaining this plugin, so this PR should be considered unsupported, and I'm certainly not keeping an eye on newly reported issues; and any maintainers aren't responsible for my mistakes (at least until they merge them).

just thinking...  I created [https://issues.jenkins.io/browse/JENKINS-71932](https://issues.jenkins.io/browse/JENKINS-71932) not realizing it was related to this change -- and now it's fixed so just wondering best way to mark [https://issues.jenkins.io/browse/JENKINS-71932](https://issues.jenkins.io/browse/JENKINS-71932) resolved?

**Comment by fengerzh** [comment]
@GLundh Could you please take a look at this PR? I know it's rude to ask someone to maintain an open source PR, I am very sorry. But this PR is really very important, as it blocks all further upgrade and keeps warn about security issue. If this PR could be approved and merged into master, it will solve many issues. 
image

**Comment by Sudarshan-TN** [comment]
> @GLundh Could you please take a look at this PR? I know it's rude to ask someone to maintain an open source PR, I am very sorry. But this PR is really very important, as it blocks all further upgrade and keeps warn about security issue. If this PR could be approved and merged into master, it will solve many issues. image

Can someone please review this PR...

**Comment by GLundh** [comment]
LGTM.

**Comment by morganchristiansson** [comment]
Now we need to push out a new release..

**Comment by emericcolombe** [comment]
Hi guys, huge thanks for this awesome work ! @GLundh do you know when a new release will be available so the plugin can get rid of its CSRF vulnerability tag ?
I'd be happy to help if necessary

**Comment by GLundh** [comment]
Not sure how automated releases work. I did not carry out the last release: https://github.com/jenkinsci/rebuild-plugin/pull/136
If you want to maintain this plugin, I'll be happy to add you to the maintainers list.

**Comment by emericcolombe** [comment]
I'm afraid I'm not really familiar with the automated release system either.. @timja could you enlighten us with your knowledge on what to do to create a release ? Thanks a lot

**Comment by an-nikolaev** [comment]
[Seems](https://github.com/jenkins-infra/github-reusable-workflows/blob/2b774121d48bf1b2d184308365e9502a96dac4c6/.github/workflows/maven-cd.yml#L17) that there should be label on PR, from [interesting categories](https://github.com/jenkinsci/.github/blob/ce466227c534c42820a597cb8e9cac2f2334920a/.github/release-drafter.yml#L29)

**Comment by lemeurherve** [comment]
> Apparently, the current version of the plugin is still being listed as being affected by the issue fixed in this PR

The warning comes from https://github.com/jenkins-infra/update-center2/blob/ed6bf6344f5b172864b4cceea788fc169bdb12f8/resources/warnings.json#L16396-L16408

To remove it, a pull request similar to https://github.com/jenkins-infra/update-center2/pull/737 has to be opened and merged.

**Comment by emericcolombe** [comment]
I created the pull request
https://github.com/jenkins-infra/update-center2/pull/752

**Comment by sghill** [comment]
Hi folks, I opened #158 that adds a null check after getting a NPE when upgrading to this version. In our case we needed to pin the plugin back to v320 to move forward.

**Pull request merged**